### ISSUE TITLE
MIDI support

### ISSUE THREAD

**Issue by tommitytom** [munshkr/nes-serial#1]
Hi there!  I'm curious how far you got with this.  I notice the MIDI references in the arduino code, but not much in the corresponding NES code.  I've been wanting to work on something like this for a while so I'm not tied to my MIDINes.  I noticed Batsley Adams was working on something similar, and apparently the MIDINes was initially written to work this way, and I'm curious as to why they stopped!

**Comment by munshkr** [comment]
Hey! I don't remember much about this, but I do remember that I ended up using a 4021 8bit shifter between the Arduino and the NES (the same IC a NES controller uses). All of the 8 output pins of the 4021 would end up connected to 8 digital pins of the Arduino Uno. That worked OK I think, probably better if I made it into a PCB (last time I played with it on a protoboard and it was a bit flimsy). It seems the README is a bit outdated, but you can see the progress in the commit log heh.

I also wondered why the MIDINes and other similar projects put an Arduino inside a cartdrige instead of trying to read from the controller port, which is, in a way, easier I think. Maybe it's too slow and would cause latency when sending MIDI commands. I never got to the point of trying that unfortunately, mainly because of a lack of 6502 skills at the time, and motivation.

If you happen to get something working let me know!

**Comment by tommitytom** [comment]
Great, thanks for the update!  Not sure when I'll have time to tackle something like this but this is where I will start :)  Also interested in debugging options - is there an emulator that lets you send data/blobs to the controller port?  Or did you do it all with a dev cart?

**Comment by orchetect** [comment]
I'm also keenly interested in this. I've written bare-metal MIDI code on various platforms and have probably just enough experience to be dangerous with Arduino and NES dev. What could possibly go wrong?

**Comment by orchetect** [comment]
I'm looking for a method without relying on a shift register. I've seen it done with just an Arduino board wired directly to the NES controller port so I know it's possible.

**Comment by munshkr** [comment]
> I'm looking for a method without relying on a shift register. I've seen it done with just an Arduino board wired directly to the NES controller port so I know it's possible.

Yes I did tried that. I switched to using the shifter because I felt it was easier than getting the timing right with the NES CPU, and I had it lying around. But it is definitely doable.

>  Also interested in debugging options - is there an emulator that lets you send data/blobs to the controller port? Or did you do it all with a dev cart?

That is something to consider. I don't have a dev cart, so I thought of extending an emulator for debugging this, but never got to that step unfortunately.
